File: packages/typescript-sdk/src/intents.ts
/**
 * Builders for intent {@link RuntimeCommand} and {@link CommandEnvelope} values.
 *
 * Pass envelopes to {@link BrowserRuntimeClient.submit}. Prefer these helpers
 * over hand-rolling nested `{ kind, input }` shapes.
 *
 * Nested wire shape:
 * `{ kind: "intent", input: { kind: "locate", input: { … } } }`.
 */
import {
  DEFAULT_DETECT_CHALLENGE_TIMEOUT_MS,
  DEFAULT_DISMISS_OBSTRUCTION_TIMEOUT_MS,
  DEFAULT_SOLVE_CHALLENGE_TIMEOUT_MS,
  MAX_INTENT_PURPOSE_BYTES,
  type CommandEnvelope,
  type CompleteFormIntent,
  type AccessibilityTarget,
  type ControlAction,
  type DetectChallengeHints,
  type DetectChallengeIntent,
  type DismissObstructionIntent,
  type ExtractField,
  type ExtractIntent,
  type FillIntent,
  type FollowIntent,
  type Id,
  type IntentHints,
  type LocateIntent,
  type RuntimeCommand,
  type SolveChallengeHints,
  type SolveChallengeIntent,
  type SubmitAndVerifyIntent,
  type WaitCondition,
  type WaitForStateIntent,
} from "./contracts.js";

/** Identifiers and deadline shared by intent envelope helpers. */
export interface IntentEnvelopeMeta {
  commandId: Id;
  workflowId: Id;
  attemptId: Id;
  sessionId: Id;
  pageId?: Id | null;
  /** ISO-8601 deadline stamped on the envelope. */
  deadline: string;
  /** Envelope schema version (default `2`). */
  schemaVersion?: number;
}

function utf8ByteLength(value: string): number {
  return new TextEncoder().encode(value).length;
}

/** Throws if `purpose` is empty or exceeds {@link MAX_INTENT_PURPOSE_BYTES}. */
export function assertIntentPurpose(purpose: string): void {
  if (purpose.length === 0) {
    throw new Error("intent purpose must be non-empty");
  }
  if (utf8ByteLength(purpose) > MAX_INTENT_PURPOSE_BYTES) {
    throw new Error(`intent purpose exceeds ${MAX_INTENT_PURPOSE_BYTES} bytes`);
  }
}

function defaultHints(): IntentHints {
  return {
    role: null,
    accessibleName: null,
    nearText: null,
    ordinal: null,
    framePath: [],
    shadowPath: [],
    allowBestMatch: false,
  };
}

/** Map an accessibility target into intent hints (`role` + exact `nearText`). */
export function intentHintsFromAccessibilityTarget(target: AccessibilityTarget): IntentHints {
  return {
    role: target.role,
    nearText: { kind: "exact", value: target.accessibleName },
    ...(target.ordinal === undefined ? {} : { ordinal: target.ordinal }),
  };
}

function withHints(hints?: IntentHints): IntentHints {
  return { ...defaultHints(), ...hints };
}

/** Build a `locate` intent command. */
export function locateRuntimeCommand(input: LocateIntent): RuntimeCommand {
  assertIntentPurpose(input.purpose);
  return {
    kind: "intent",
    input: {
      kind: "locate",
      input: {
        purpose: input.purpose,
        hints: withHints(input.hints),
      },
    },
  };
}

/** Build a `fill` intent command. */
export function fillRuntimeCommand(input: FillIntent): RuntimeCommand {
  assertIntentPurpose(input.purpose);
  return {
    kind: "intent",
    input: {
      kind: "fill",
      input: {
        purpose: input.purpose,
        hints: withHints(input.hints),
        value: normalizeFillValue(input.value),
      },
    },
  };
}

/** Build a `completeForm` intent command (1–128 uniquely named fields). */
export function completeFormRuntimeCommand(input: CompleteFormIntent): RuntimeCommand {
  assertIntentPurpose(input.purpose);
  if (input.fields.length === 0) throw new Error("completeForm fields must not be empty");
  if (input.fields.length > 128) throw new Error("completeForm fields must not exceed 128 items");
  const names = new Set<string>();
  const fields = input.fields.map((field) => {
    if (field.name.trim().length === 0) throw new Error("completeForm field name must not be empty");
    if (names.has(field.name)) throw new Error(`duplicate completeForm field name: ${field.name}`);
    names.add(field.name);
    assertIntentPurpose(field.purpose);
    return { ...field, hints: withHints(field.hints), value: normalizeFillValue(field.value) };
  });
  return { kind: "intent", input: { kind: "completeForm", input: { purpose: input.purpose, fields } } };
}

/** Emits the canonical wire shape; `setText` without `clearFirst` means replace. */
function normalizeFillValue(value: ControlAction): ControlAction {
  if (value.kind === "setText") {
    return { kind: "setText", value: value.value, clearFirst: value.clearFirst ?? true };
  }
  return value;
}

/** Build a `submitAndVerify` intent command. */
export function submitAndVerifyRuntimeCommand(input: SubmitAndVerifyIntent): RuntimeCommand {
  assertIntentPurpose(input.purpose);
  return {
    kind: "intent",
    input: {
      kind: "submitAndVerify",
      input: {
        purpose: input.purpose,
        hints: withHints(input.hints),
        expectedState: input.expectedState,
      },
    },
  };
}

/** Build a `waitForState` intent command. */
export function waitForStateRuntimeCommand(input: WaitForStateIntent): RuntimeCommand {
  return {
    kind: "intent",
    input: {
      kind: "waitForState",
      input: {
        condition: input.condition,
        timeoutMs: input.timeoutMs,
      },
    },
  };
}

/** Build a `follow` intent command. */
export function followRuntimeCommand(input: FollowIntent): RuntimeCommand {
  assertIntentPurpose(input.purpose);
  return {
    kind: "intent",
    input: {
      kind: "follow",
      input: {
        purpose: input.purpose,
        hints: withHints(input.hints),
        expectedDestination: input.expectedDestination,
        boundary: input.boundary ?? false,
      },
    },
  };
}

/** Build a `dismissObstruction` intent command. */
export function dismissObstructionRuntimeCommand(input: DismissObstructionIntent): RuntimeCommand {
  assertIntentPurpose(input.purpose);
  return {
    kind: "intent",
    input: {
      kind: "dismissObstruction",
      input: {
        purpose: input.purpose,
        hints: withHints(input.hints),
        timeoutMs: input.timeoutMs ?? DEFAULT_DISMISS_OBSTRUCTION_TIMEOUT_MS,
      },
    },
  };
}

/** Build an `extract` intent command (at least one uniquely named field). */
export function extractRuntimeCommand(input: ExtractIntent): RuntimeCommand {
  assertIntentPurpose(input.purpose);
  if (input.fields.length === 0) {
    throw new Error("extract intent must include at least one field");
  }
  const seenNames = new Set<string>();
  const fields: ExtractField[] = input.fields.map((field) => {
    assertIntentPurpose(field.purpose);
    const name = field.name.trim();
    if (name.length === 0) {
      throw new Error("extract field name must not be empty");
    }
    if (seenNames.has(name)) {
      throw new Error(`duplicate extract field name: ${name}`);
    }
    seenNames.add(name);
    return {
      name,
      purpose: field.purpose,
      hints: withHints(field.hints),
      value: field.value,
    };
  });
  return {
    kind: "intent",
    input: {
      kind: "extract",
      input: {
        purpose: input.purpose,
        fields,
      },
    },
  };
}

/** Wrap an intent {@link RuntimeCommand} in a {@link CommandEnvelope}. */
export function intentEnvelope(meta: IntentEnvelopeMeta, command: RuntimeCommand): CommandEnvelope {
  if (command.kind !== "intent") {
    throw new Error('intentEnvelope requires RuntimeCommand with kind "intent"');
  }
  return {
    schemaVersion: meta.schemaVersion ?? 2,
    commandId: meta.commandId,
    workflowId: meta.workflowId,
    attemptId: meta.attemptId,
    sessionId: meta.sessionId,
    pageId: meta.pageId ?? null,
    deadline: meta.deadline,
    command,
  };
}

/** Convenience: {@link locateRuntimeCommand} + {@link intentEnvelope}. */
export function locateEnvelope(meta: IntentEnvelopeMeta, purpose: string, hints?: IntentHints): CommandEnvelope {
  return intentEnvelope(meta, locateRuntimeCommand({ purpose, hints }));
}

/** Convenience: {@link fillRuntimeCommand} + {@link intentEnvelope}. */
export function fillEnvelope(meta: IntentEnvelopeMeta, purpose: string, value: ControlAction, hints?: IntentHints): CommandEnvelope {
  return intentEnvelope(meta, fillRuntimeCommand({ purpose, value, hints }));
}

/** Convenience: {@link submitAndVerifyRuntimeCommand} + {@link intentEnvelope}. */
export function submitAndVerifyEnvelope(
  meta: IntentEnvelopeMeta,
  purpose: string,
  expectedState: SubmitAndVerifyIntent["expectedState"],
  hints?: IntentHints,
): CommandEnvelope {
  return intentEnvelope(meta, submitAndVerifyRuntimeCommand({ purpose, expectedState, hints }));
}

/** Convenience: {@link waitForStateRuntimeCommand} + {@link intentEnvelope}. */
export function waitForStateEnvelope(
  meta: IntentEnvelopeMeta,
  condition: WaitCondition,
  timeoutMs: number,
): CommandEnvelope {
  return intentEnvelope(meta, waitForStateRuntimeCommand({ condition, timeoutMs }));
}

/** Convenience: {@link followRuntimeCommand} + {@link intentEnvelope}. */
export function followEnvelope(
  meta: IntentEnvelopeMeta,
  purpose: string,
  expectedDestination: FollowIntent["expectedDestination"],
  options?: { hints?: IntentHints; boundary?: boolean },
): CommandEnvelope {
  return intentEnvelope(
    meta,
    followRuntimeCommand({
      purpose,
      expectedDestination,
      hints: options?.hints,
      boundary: options?.boundary,
    }),
  );
}

/** Convenience: {@link dismissObstructionRuntimeCommand} + {@link intentEnvelope}. */
export function dismissObstructionEnvelope(
  meta: IntentEnvelopeMeta,
  purpose: string,
  options?: { hints?: IntentHints; timeoutMs?: number },
): CommandEnvelope {
  return intentEnvelope(
    meta,
    dismissObstructionRuntimeCommand({
      purpose,
      hints: options?.hints,
      timeoutMs: options?.timeoutMs,
    }),
  );
}

/** Convenience: {@link extractRuntimeCommand} + {@link intentEnvelope}. */
export function extractEnvelope(
  meta: IntentEnvelopeMeta,
  purpose: string,
  fields: ExtractField[],
): CommandEnvelope {
  return intentEnvelope(meta, extractRuntimeCommand({ purpose, fields }));
}

/**
 * Build a `detectChallenge` intent command (Replayable, read-only):
 * screenshot in, `challengeDetection` evidence out — a provably clean page
 * is a first-class answer. Never acts on the page.
 */
export function detectChallengeRuntimeCommand(input: DetectChallengeIntent): RuntimeCommand {
  assertIntentPurpose(input.purpose);
  const hints: DetectChallengeHints = {
    timeoutMs: input.hints?.timeoutMs ?? DEFAULT_DETECT_CHALLENGE_TIMEOUT_MS,
    ...(input.hints?.region ? { region: input.hints.region } : {}),
  };
  return {
    kind: "intent",
    input: { kind: "detectChallenge", input: { purpose: input.purpose, hints } },
  };
}

/** Build a `solveChallenge` intent command (Reconciliable vision solve loop). */
export function solveChallengeRuntimeCommand(input: SolveChallengeIntent): RuntimeCommand {
  assertIntentPurpose(input.purpose);
  const hints: SolveChallengeHints = {
    timeoutMs: input.hints?.timeoutMs ?? DEFAULT_SOLVE_CHALLENGE_TIMEOUT_MS,
    ...(input.hints?.region ? { region: input.hints.region } : {}),
  };
  return {
    kind: "intent",
    input: { kind: "solveChallenge", input: { purpose: input.purpose, hints } },
  };
}

/** Convenience: {@link detectChallengeRuntimeCommand} + {@link intentEnvelope}. */
export function detectChallengeEnvelope(
  meta: IntentEnvelopeMeta,
  purpose: string,
  options?: { hints?: DetectChallengeHints },
): CommandEnvelope {
  return intentEnvelope(
    meta,
    detectChallengeRuntimeCommand({ purpose, hints: options?.hints }),
  );
}

/** Convenience: {@link solveChallengeRuntimeCommand} + {@link intentEnvelope}. */
export function solveChallengeEnvelope(
  meta: IntentEnvelopeMeta,
  purpose: string,
  options?: { hints?: SolveChallengeHints },
): CommandEnvelope {
  return intentEnvelope(
    meta,
    solveChallengeRuntimeCommand({ purpose, hints: options?.hints }),
  );
}

File: crates/bobby-browser-client/src/intents.rs
//! Builders for intent [`RuntimeCommand`] and [`CommandEnvelope`] values —
//! the Rust twin of the TypeScript SDK's `intents` module.
//!
//! Pass envelopes to [`BrowserRuntimeClient::submit`](crate::http::BrowserRuntimeClient::submit).
//! Prefer these helpers over hand-rolling nested `{ kind, input }` shapes.
//!
//! Nested wire shape:
//! `{ kind: "intent", input: { kind: "locate", input: { … } } }`.

use crate::challenges::{DetectChallengeIntent, SolveChallengeIntent};
use crate::commands::{
    CommandEnvelope, CompleteFormIntent, ControlAction, DismissObstructionIntent, ExtractField,
    ExtractIntent, FillIntent, FollowIntent, IntentCommand, IntentHints, LocateIntent,
    RuntimeCommand, SubmitAndVerifyIntent, WaitCondition, WaitForCommand, WaitForStateIntent,
    MAX_INTENT_PURPOSE_BYTES,
};
use crate::ids::{AttemptId, CommandId, PageId, SessionId, WorkflowId};
use chrono::{DateTime, Utc};

/// Meta that every intent envelope carries. Mirrors the TypeScript SDK's
/// `IntentEnvelopeMeta`.
#[derive(Debug, Clone)]
pub struct IntentEnvelopeMeta {
    pub command_id: CommandId,
    pub workflow_id: WorkflowId,
    pub attempt_id: AttemptId,
    pub session_id: SessionId,
    pub page_id: Option<PageId>,
    /// RFC-3339 deadline stamped on the envelope.
    pub deadline: DateTime<Utc>,
}

/// Throws the Rust equivalent of the TS SDK's `assertIntentPurpose`: an
/// empty or over-budget purpose fails the builder instead of failing at the
/// runtime after a round trip.
fn assert_intent_purpose(purpose: &str) {
    if purpose.len() > MAX_INTENT_PURPOSE_BYTES {
        panic!(
            "intent purpose exceeds {MAX_INTENT_PURPOSE_BYTES} bytes (got {})",
            purpose.len()
        );
    }
    if purpose.trim().is_empty() {
        panic!("intent purpose must be a non-empty string");
    }
}

/// `setText`'s `clear_first` is a plain `bool` with a serde default on the
/// wire, so Rust callers get the TS SDK's normalization for free: an absent
/// `clearFirst` decodes to `true` (replace). The builder is an identity
/// function kept for parity with the TypeScript module.
fn normalize_fill_value(value: ControlAction) -> ControlAction {
    value
}

/// Wrap an intent [`RuntimeCommand`] in a [`CommandEnvelope`].
pub fn intent_envelope(meta: IntentEnvelopeMeta, command: RuntimeCommand) -> CommandEnvelope {
    debug_assert!(
        matches!(command, RuntimeCommand::Intent(_)),
        "intentEnvelope requires a RuntimeCommand with kind \"intent\""
    );
    CommandEnvelope {
        schema_version: CommandEnvelope::SCHEMA_VERSION,
        command_id: meta.command_id,
        workflow_id: meta.workflow_id,
        attempt_id: meta.attempt_id,
        session_id: meta.session_id,
        page_id: meta.page_id,
        deadline: meta.deadline,
        command,
    }
}

/// Build a `locate` intent command.
pub fn locate_runtime_command(input: LocateIntent) -> RuntimeCommand {
    assert_intent_purpose(&input.purpose);
    RuntimeCommand::Intent(IntentCommand::Locate(LocateIntent {
        purpose: input.purpose,
        hints: input.hints,
    }))
}

/// Build a `fill` intent command.
pub fn fill_runtime_command(input: FillIntent) -> RuntimeCommand {
    assert_intent_purpose(&input.purpose);
    RuntimeCommand::Intent(IntentCommand::Fill(FillIntent {
        purpose: input.purpose,
        hints: input.hints,
        value: normalize_fill_value(input.value),
    }))
}

/// Build a `completeForm` intent command (1–128 uniquely named fields).
pub fn complete_form_runtime_command(input: CompleteFormIntent) -> RuntimeCommand {
    assert_intent_purpose(&input.purpose);
    assert!(
        !input.fields.is_empty(),
        "completeForm fields must not be empty"
    );
    assert!(
        input.fields.len() <= 128,
        "completeForm fields must not exceed 128 items"
    );
    let mut names = std::collections::HashSet::new();
    let fields = input
        .fields
        .into_iter()
        .map(|mut field| {
            assert!(
                !field.name.trim().is_empty(),
                "completeForm field name must not be empty"
            );
            assert!(
                names.insert(field.name.clone()),
                "duplicate completeForm field name: {}",
                field.name
            );
            assert_intent_purpose(&field.purpose);
            field.value = normalize_fill_value(field.value);
            field
        })
        .collect();
    RuntimeCommand::Intent(IntentCommand::CompleteForm(CompleteFormIntent {
        purpose: input.purpose,
        fields,
    }))
}

/// Build a `submitAndVerify` intent command.
pub fn submit_and_verify_runtime_command(input: SubmitAndVerifyIntent) -> RuntimeCommand {
    assert_intent_purpose(&input.purpose);
    RuntimeCommand::Intent(IntentCommand::SubmitAndVerify(SubmitAndVerifyIntent {
        purpose: input.purpose,
        hints: input.hints,
        expected_state: input.expected_state,
    }))
}

/// Build a `waitForState` intent command.
pub fn wait_for_state_runtime_command(input: WaitForStateIntent) -> RuntimeCommand {
    RuntimeCommand::Intent(IntentCommand::WaitForState(WaitForStateIntent {
        condition: input.condition,
        timeout_ms: input.timeout_ms,
    }))
}

/// Build a `follow` intent command.
pub fn follow_runtime_command(input: FollowIntent) -> RuntimeCommand {
    assert_intent_purpose(&input.purpose);
    RuntimeCommand::Intent(IntentCommand::Follow(FollowIntent {
        purpose: input.purpose,
        hints: input.hints,
        expected_destination: input.expected_destination,
        boundary: input.boundary,
    }))
}

/// Build a `dismissObstruction` intent command.
pub fn dismiss_obstruction_runtime_command(input: DismissObstructionIntent) -> RuntimeCommand {
    assert_intent_purpose(&input.purpose);
    RuntimeCommand::Intent(IntentCommand::DismissObstruction(
        DismissObstructionIntent {
            purpose: input.purpose,
            hints: input.hints,
            timeout_ms: input.timeout_ms,
        },
    ))
}

/// Build an `extract` intent command (at least one uniquely named field).
pub fn extract_runtime_command(input: ExtractIntent) -> RuntimeCommand {
    assert_intent_purpose(&input.purpose);
    assert!(
        !input.fields.is_empty(),
        "extract intent must include at least one field"
    );
    let mut seen = std::collections::HashSet::new();
    for field in &input.fields {
        assert_intent_purpose(&field.purpose);
        assert!(
            seen.insert(field.name.clone()),
            "duplicate extract field name: {}",
            field.name
        );
    }
    RuntimeCommand::Intent(IntentCommand::Extract(ExtractIntent {
        purpose: input.purpose,
        fields: input.fields,
    }))
}

/// Build a `detectChallenge` intent command (Replayable, read-only).
pub fn detect_challenge_runtime_command(input: DetectChallengeIntent) -> RuntimeCommand {
    assert_intent_purpose(&input.purpose);
    RuntimeCommand::Intent(IntentCommand::DetectChallenge(DetectChallengeIntent {
        purpose: input.purpose,
        hints: input.hints,
    }))
}

/// Build a `solveChallenge` intent command (Reconciliable vision solve loop).
pub fn solve_challenge_runtime_command(input: SolveChallengeIntent) -> RuntimeCommand {
    assert_intent_purpose(&input.purpose);
    RuntimeCommand::Intent(IntentCommand::SolveChallenge(SolveChallengeIntent {
        purpose: input.purpose,
        hints: input.hints,
    }))
}

/// Convenience: [`locate`] runtime command + [`intent_envelope`].
pub fn locate_envelope(
    meta: IntentEnvelopeMeta,
    purpose: &str,
    hints: Option<IntentHints>,
) -> CommandEnvelope {
    intent_envelope(
        meta,
        locate_runtime_command(LocateIntent {
            purpose: purpose.into(),
            hints: hints.unwrap_or_default(),
        }),
    )
}

/// Convenience: [`fill`] runtime command + [`intent_envelope`].
pub fn fill_envelope(
    meta: IntentEnvelopeMeta,
    purpose: &str,
    value: ControlAction,
    hints: Option<IntentHints>,
) -> CommandEnvelope {
    intent_envelope(
        meta,
        fill_runtime_command(FillIntent {
            purpose: purpose.into(),
            value,
            hints: hints.unwrap_or_default(),
        }),
    )
}

/// Convenience: [`submit_and_verify`] runtime command + [`intent_envelope`].
pub fn submit_and_verify_envelope(
    meta: IntentEnvelopeMeta,
    purpose: &str,
    expected_state: WaitForCommand,
    hints: Option<IntentHints>,
) -> CommandEnvelope {
    intent_envelope(
        meta,
        submit_and_verify_runtime_command(SubmitAndVerifyIntent {
            purpose: purpose.into(),
            expected_state,
            hints: hints.unwrap_or_default(),
        }),
    )
}

/// Convenience: [`wait_for_state`] runtime command + [`intent_envelope`].
pub fn wait_for_state_envelope(
    meta: IntentEnvelopeMeta,
    condition: WaitCondition,
    timeout_ms: u64,
) -> CommandEnvelope {
    intent_envelope(
        meta,
        wait_for_state_runtime_command(WaitForStateIntent {
            condition,
            timeout_ms,
        }),
    )
}

/// Convenience: [`follow`] runtime command + [`intent_envelope`].
pub fn follow_envelope(
    meta: IntentEnvelopeMeta,
    purpose: &str,
    expected_destination: WaitForCommand,
    boundary: bool,
    hints: Option<IntentHints>,
) -> CommandEnvelope {
    intent_envelope(
        meta,
        follow_runtime_command(FollowIntent {
            purpose: purpose.into(),
            expected_destination,
            hints: hints.unwrap_or_default(),
            boundary,
        }),
    )
}

/// Convenience: [`dismiss_obstruction`] runtime command + [`intent_envelope`].
pub fn dismiss_obstruction_envelope(
    meta: IntentEnvelopeMeta,
    purpose: &str,
    timeout_ms: Option<u64>,
) -> CommandEnvelope {
    intent_envelope(
        meta,
        dismiss_obstruction_runtime_command(DismissObstructionIntent {
            purpose: purpose.into(),
            hints: IntentHints::default(),
            timeout_ms: timeout_ms
                .unwrap_or(crate::commands::DEFAULT_DISMISS_OBSTRUCTION_TIMEOUT_MS),
        }),
    )
}

/// Convenience: [`extract`] runtime command + [`intent_envelope`].
pub fn extract_envelope(
    meta: IntentEnvelopeMeta,
    purpose: &str,
    fields: Vec<ExtractField>,
) -> CommandEnvelope {
    intent_envelope(
        meta,
        extract_runtime_command(ExtractIntent {
            purpose: purpose.into(),
            fields,
        }),
    )
}

/// Convenience: [`detect_challenge`] runtime command + [`intent_envelope`].
pub fn detect_challenge_envelope(
    meta: IntentEnvelopeMeta,
    purpose: &str,
    hints: Option<crate::challenges::DetectChallengeHints>,
) -> CommandEnvelope {
    intent_envelope(
        meta,
        detect_challenge_runtime_command(DetectChallengeIntent {
            purpose: purpose.into(),
            hints: hints.unwrap_or_default(),
        }),
    )
}

/// Convenience: [`solve_challenge`] runtime command + [`intent_envelope`].
pub fn solve_challenge_envelope(
    meta: IntentEnvelopeMeta,
    purpose: &str,
    hints: Option<crate::challenges::SolveChallengeHints>,
) -> CommandEnvelope {
    intent_envelope(
        meta,
        solve_challenge_runtime_command(SolveChallengeIntent {
            purpose: purpose.into(),
            hints: hints.unwrap_or_default(),
        }),
    )
}
#[cfg(test)]
mod tests {
    use super::*;
    use crate::commands::{CompleteFormField, ExtractValueKind, WaitUntil};
    use crate::ids::PageId;

    fn meta() -> IntentEnvelopeMeta {
        IntentEnvelopeMeta {
            command_id: CommandId::new(),
            workflow_id: WorkflowId::new(),
            attempt_id: AttemptId::new(),
            session_id: SessionId::new(),
            page_id: Some(PageId::new()),
            deadline: Utc::now(),
        }
    }

    /// Golden wire shape: the nested `{kind, input}` envelope matches the
    /// TypeScript SDK's byte-for-byte (serde tag + camelCase rename).
    #[test]
    fn submit_envelope_wraps_the_nested_wire_shape() {
        let envelope = submit_and_verify_envelope(
            meta(),
            "Submit the customer priority",
            WaitForCommand {
                condition: WaitCondition::Document {
                    ready: WaitUntil::Interactive,
                },
                timeout_ms: 1_000,
            },
            None,
        );
        assert_eq!(envelope.schema_version, CommandEnvelope::SCHEMA_VERSION);
        match &envelope.command {
            RuntimeCommand::Intent(IntentCommand::SubmitAndVerify(submit)) => {
                assert_eq!(submit.purpose, "Submit the customer priority");
                assert_eq!(
                    submit.expected_state.condition,
                    WaitCondition::Document {
                        ready: WaitUntil::Interactive
                    }
                );
            }
            other => panic!("expected intent submitAndVerify, got {other:?}"),
        }
    }

    #[test]
    fn complete_form_builds_unique_named_fields() {
        let command = complete_form_runtime_command(CompleteFormIntent {
            purpose: "fill the onboarding form".into(),
            fields: vec![CompleteFormField {
                name: "email".into(),
                purpose: "work address".into(),
                hints: IntentHints::default(),
                value: ControlAction::SetText {
                    value: "ada@example.test".into(),
                    clear_first: true,
                },
            }],
        });
        match command {
            RuntimeCommand::Intent(IntentCommand::CompleteForm(form)) => {
                assert_eq!(form.fields.len(), 1);
            }
            other => panic!("expected completeForm, got {other:?}"),
        }
    }

    #[test]
    fn duplicate_extract_field_names_are_refused() {
        let field = ExtractField {
            name: "email".into(),
            purpose: "address".into(),
            hints: IntentHints::default(),
            value: crate::commands::ExtractValueKind::Text,
        };
        let result = std::panic::catch_unwind(|| {
            extract_runtime_command(ExtractIntent {
                purpose: "read the header".into(),
                fields: vec![field.clone(), field],
            })
        });
        assert!(result.is_err(), "duplicate field names must panic");
    }

    #[test]
    fn challenge_envelopes_carry_their_hints() {
        let envelope = solve_challenge_envelope(
            meta(),
            "clear the recaptcha",
            Some(crate::challenges::SolveChallengeHints {
                region: Some(crate::challenges::ChallengeRegion {
                    x: 1.0,
                    y: 2.0,
                    width: 3.0,
                    height: 4.0,
                }),
                timeout_ms: 45_000,
            }),
        );
        match &envelope.command {
            RuntimeCommand::Intent(IntentCommand::SolveChallenge(solve)) => {
                assert_eq!(solve.hints.timeout_ms, 45_000);
                assert!(solve.hints.region.is_some());
            }
            other => panic!("expected solveChallenge, got {other:?}"),
        }
    }
}
